Add Rowwise Prune PyTorch op

radkris-git opened the pull request:
Summary:
Add rowwise prune pytorch op.

This operator introduces sparsity to the 'weights' matrix with the help
of the importance indicator 'mask'.

A row is considered important and not pruned if the mask value for that
particular row is 1(True) and not important otherwise.

Test Plan:
buck test caffe2/torch/fb/sparsenn:test -- rowwise_prune
buck test caffe2/test:pruning

Differential Revision: D22849432

raghuramank100 commented on aten/src/ATen/native/RowwisePrune.cpp:
```diff
@@ -0,0 +1,101 @@
+  int last_row_kept = 0;
+  for (int i = 0; i < mask.numel(); i++) {
+    if (mask_data[i]) {
+      memcpy(pruned_2d_tensor_data + last_row_kept * num_cols,
```
Does this work for quantized weights?

radkris-git replied:
Good point. Made it generic to work all integral + float types using AT_DISPATCH_ALL_TYPES.

raghuramank100 commented on test/test_pruning_op.py:
```diff
@@ -0,0 +1,56 @@
+
+class PruningOpTest(TestCase):
+
+    def _test_rowwise_prune_op(self, embedding_rows, embedding_dims, indices_type):
```
Please add a test to check if this op is scriptable. Since it takes a dtype as an argument, I am not sure if it is

radkris-git replied:
I think this comment doesn't apply as it's just a pytorch c++ op and not a nn.Module.

raghuramank100 commented on aten/src/ATen/native/RowwisePrune.cpp:
```diff
@@ -0,0 +1,101 @@
+      compressed_indices_dtype);
+  auto compressed_indices_mapping_data =
+      compressed_indices_mapping.data_ptr<input_t>();
+  auto weights_data = weights.data_ptr<float>();
```
If this op works only for floating point weights, lets check the dtype of the tensor

radkris-git replied:
Good point. Made it generic with AT_DISPATCH_ALL_TYPES.

raghuramank100 reviewed (changes requested):
Please see comments

supriyar commented on test/test_pruning_op.py:
```diff
@@ -0,0 +1,56 @@
+
+    def _test_rowwise_prune_op(self, embedding_rows, embedding_dims, indices_type):
+        embedding_weights = torch.from_numpy((np.random.random_sample((embedding_rows, embedding_dims)) + 1).astype(np.float32))
+        mask = torch.from_numpy(np.random.choice(
```
Should we modify the test to show how the mask can be generated given indicator and threshold values (which can be random values in the test)?

radkris-git replied:
SG. Done

radkris-git commented:
Addressed comments.

supriyar commented on aten/src/ATen/native/RowwisePrune.cpp:
```diff
@@ -0,0 +1,103 @@
+      weights.options());
+  auto compressed_indices_mapping = at::empty({mask.numel()},
+      compressed_indices_dtype);
+  AT_DISPATCH_ALL_TYPES(weights.scalar_type(), "rowwise_prune_helper", [&]() {
```
I think this doesn't support at::Half types, can we include that too?

supriyar commented on test/test_pruning_op.py:
```diff
@@ -0,0 +1,70 @@
+        return mask
+
+    def _test_rowwise_prune_op(self, embedding_rows, embedding_dims, indices_type):
+        weight_types = [np.int8, np.int32, np.int16, np.float32, np.double]
```
Can we add this as a hypothesis argument?

supriyar replied:
Also it would be better to add torch types here rather than numpy types.

radkris-git replied:
Good points.

Commit:
Add Rowwise Prune PyTorch op (#42708)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42708

Add rowwise prune pytorch op.

This operator introduces sparsity to the 'weights' matrix with the help
of the importance indicator 'mask'.

A row is considered important and not pruned if the mask value for that
particular row is 1(True) and not important otherwise.

Test Plan:
buck test caffe2/torch/fb/sparsenn:test -- rowwise_prune
buck test caffe2/test:pruning

Reviewed By: supriyar

Differential Revision: D22849432

fbshipit-source-id: b9eaa128fd8b8d63cf1aa2270dbb23f7816bf79d
